tracks/serializers.py

```diff
@@ -1,43 +1,68 @@
-from .models import Genome, Category, Track, Source
+from .models import Category, Track, Source
 from rest_framework import serializers
 
 """
-DRF Serializers corresponding to tracks app datamodels
+Serializers for Track API datamodels
 """
 
 class SourceSerializer(serializers.ModelSerializer):
     class Meta:
         model = Source
         fields = ["name", "url"]
 
-
-class ClientTrackSerializer(serializers.ModelSerializer):
-    sources = SourceSerializer(many=True, read_only=True)
+class BaseTrackSerializer(serializers.ModelSerializer):
+    sources = SourceSerializer(many=True)
 
     class Meta:
         model = Track
-        fields = ["label", "track_id", "colour", "trigger", "type",
-            "display_order", "on_by_default", "additional_info", "description", "sources"]
+        fields = ["track_id", "label", "colour", "trigger", "type", "display_order", "on_by_default", "sources"]
 
-class GenomeBrowserTrackSerializer(serializers.ModelSerializer):
-    class Meta:
-        model = Track
-        fields = ["label", "track_id", "colour", "trigger", "type", "datafiles",
-            "display_order", "on_by_default"]
+# track payload in "track_categories" endpoint (consumed by client)
+class CategoryTrackSerializer(BaseTrackSerializer):
+    class Meta(BaseTrackSerializer.Meta):
+        fields = BaseTrackSerializer.Meta.fields + ["additional_info", "description"]
 
+# track payload in "track" endpoint (consumed by genome browser)
+class ReadTrackSerializer(BaseTrackSerializer):
+    class Meta(BaseTrackSerializer.Meta):
+        fields = BaseTrackSerializer.Meta.fields + ["datafiles"]
 
 class CategorySerializer(serializers.ModelSerializer):
-    types = serializers.StringRelatedField(many=True)
-    track_list = ClientTrackSerializer(many=True, read_only=True)
-
     class Meta:
         model = Category
-        fields = ["label", "track_category_id", "types", "track_list"]
+        fields = ["label", "track_category_id", "type"]
+        extra_kwargs = {
+            "track_category_id": {"validators": []}
+        }
+
+# track submission payload
+class WriteTrackSerializer(BaseTrackSerializer):
+    category = CategorySerializer(write_only=True)
 
+    class Meta(BaseTrackSerializer.Meta):
+        fields = BaseTrackSerializer.Meta.fields + ["genome_id", "category", "datafiles", "additional_info", "description"]
+        extra_kwargs = {
+            "genome_id": {"write_only": True},
+            "sources": {"required": False}
+        }
+    
+    def create(self, validated_data):
```

Comment:
1. Is serialiser the right place for this?
2. Do we want to get the object when POST request is sent or do we need a simple create and exceptions?

Reply:
1. `create()` is part of DRF serializers to implement writing payloads to database. Here I am overriding `create()` in order to support writable nested serializers: the incoming payload needs to be split and stored across multiple tables/datamodels ([relevant docs](https://www.django-rest-framework.org/api-guide/serializers/#writing-create-methods-for-nested-representations)).
2. get_or_create() is used here because I'm following the [database model](https://lucid.app/lucidchart/5cee5c19-5c63-4ba5-b774-7c75e8af4319/edit?viewport_loc=129%2C209%2C1928%2C948%2C0_0&invitationId=inv_56eaae93-6782-4471-a8a8-7a6a661fadee): I need to first create any new category/source entries as they come in and then I reuse+link the existing entries as relevant in subsequent requests (instead of raising an exception).